### DISCUSSION TITLE
[2024-01-01] Library & Engine Team Update

### DISCUSSION THREAD

**Discussion by GregoryTravis** [enso-org/enso#8631] [Weekly Demo Day]
- James: Minor fixes and tweak for Cloud APIs. #8557
- James: A few minor fixes from release testing today. #8562
- James: Fix Table viz error #8563
- James: Small linting fixes. #8592
- James: Add a quick support for pasting data from Excel. #8605
- Radosław: Getting pending PRs merged:
  - Improving tests and edge cases for URI and HTTP #8497
  - Opt-in return type checks #8502
  - Add tests for Enso Cloud auth + simple API mock for `Enso_User` #8511
  - Add a few new `Filter_Condition`s #8539
  - Fixes `Date_Diff` widget #8561
- Radosław: Better error message in `Type_Error` raised from return type checks 
  #8566
- Radosław: Remove the Apache dependency from `std-base` #8571
- Radosław: Merge `URI_With_Query` into `URI`, extend API of `URI` #8591
- Radosław: Handle `Nothing` values in `Filter_Condition.to_predicate` #8600
- Radosław: (almost ready) Unify `File_Format_Metadata` with `File_For_Read` #8628
- Radosław: Working on Type inference PoC #8652
  - Inferring types of literals `1 ==> Integer`, `"foo" ==> Text`
  - Inferring types of applied constructors `(My_Type.Value 1 2 3) ==> My_Type` (if there are no defaults in the constructor - defaults will be added later).
  - Inferring types of values with ascription `(foo.bar : Integer) ==> Integer`
  - Type checking is enabled by a CLI flag `--enable-type-check`, currently defaults to false.
  - First simple type error reported - detecting `Not_Invokable` error in simple cases.
  - Not yet finished: propagating type information through bindings (`x = 9` should imply `y = x` to also be Integer).
- Greg: Support on_problems=Problem_Behavior.Report_Warning and Map_Error wrapping in Vector.map #8595
- Greg: Ignore User-Agent in HTTP tests #8630
- Cassandra: Renamed lookup_and_replace to merge and renamed Table.replace to text… #8564
- Jaroslav: Avoid no WithWarnings warnings #8583
- Jaroslav: More robust against debris in the repository #8596
- Jaroslav: **7000** times faster `DataflowError` #8608
- Jaroslav: size of `Table.ir` down from 6MB to 3.5MB #8635 
- Jaroslav: Autoscoping [via unresolved symbol](https://github.com/enso-org/enso/issues/8645#issuecomment-1871791238) prototype. Join discussion: #8646
- Hubert: Arrow language #8512
- Hubert: Delaying builtins initialization to speedup startup #8654 
- Hubert: Stop parsing METADATA section for IdMap #8574
- Hubert: Ignore spurious warnings #8575
- Hubert: Add thread dump info to investigate MacOS timeouts #8577
- Hubert: Carry position for project management operations #8589
- Pavel: Update java formatter sbt plugin #8543
- Pavel: Submodules can be private #8581
- Pavel: All unit tests use the same testing logging provider #8593
- Pavel: Fix build of engine benchmarks #8620
- Pavel: Refactor Standard.Test to the builder API #8622 
- Pavel: Increase warmup for `EqualsBenchmarks` #8636 
- Pavel: Format Java sources in benchmark configuration #8638
- Dmitry: Fix serialization of alias analysis graph #8550
- Dmitry: Syntax error results in silent error #8560
- Dmitry: Runtime test build depends on fat jar classes #8588
- Dmitry: Unable to access constructor value in interactive mode #8626

**Comment by radeusgd** [comment]
```
foo x =
    if x == 123 then 1 2 else 123

main =
    foo 10
```

run with `--enable-type-check` yields:

![image](https://github.com/enso-org/enso/assets/1436948/65910375-9f1a-40b3-af38-cc38e7d1269e)
